apps/HelloAndroidCamera2/jni/BufferTFunctions.cpp

```diff
@@ -0,0 +1,147 @@
+#include "BufferTFunctions.h"
+
+#include <algorithm>
+#include <string.h>
+
+#include <android/log.h>
+
+#define LOGD(...) __android_log_print(ANDROID_LOG_DEBUG, "BufferTFunctions", __VA_ARGS__)
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR, "BufferTFunctions", __VA_ARGS__)
+
+bool isHostNull(const buffer_t &buf) {
+    return (buf.host == nullptr);
+}
+
+bool equalExtents(const buffer_t &a, const buffer_t &b) {
+    return (a.extent[0] == b.extent[0] &&
+        a.extent[1] == b.extent[1] &&
+        a.extent[2] == b.extent[2] &&
+        a.extent[3] == b.extent[3]);
+}
+
+bool equalStrides(const buffer_t &a, const buffer_t &b) {
+    return (a.stride[0] == b.stride[0] &&
+        a.stride[1] == b.stride[1] &&
+        a.stride[2] == b.stride[2] &&
+        a.stride[3] == b.stride[3]);
+}
+
+bool copy2D(const buffer_t &src, const buffer_t &dst) {
+    if (!equalExtents(src, dst)) {
+        return false;
+    }
+    if (src.elem_size != dst.elem_size) {
+        return false;
+    }
+
+    if (src.stride[0] == 1 && src.stride[1] == src.extent[0] &&
+        dst.stride[0] == 1 && dst.stride[1] == dst.extent[0]) {
+        // Copy all at once.
+        memcpy(dst.host, src.host, src.extent[1] * src.stride[1] * src.elem_size);
+    } else if (src.stride[0] == 1 && dst.stride[0] == 1) {
+        // memcpy row by row.
+        int32_t minRowStrideBytes = std::min(src.stride[1], dst.stride[1]) * src.elem_size;
+        for (int32_t y = 0; y < src.extent[1]; ++y) {
+            const uint8_t *srcRow = src.host + y * src.stride[1] * src.elem_size;
+            uint8_t *dstRow = dst.host + y * dst.stride[1] * dst.elem_size;
+            memcpy(dstRow, srcRow, minRowStrideBytes);
+        }
+    } else {
+        // Copy element by element.
+        int32_t srcElementStrideBytes = src.stride[0] * src.elem_size;
+        int32_t dstElementStrideBytes = dst.stride[0] * dst.elem_size;
+        int32_t srcRowStrideBytes = src.stride[1] * src.elem_size;
+        int32_t dstRowStrideBytes = dst.stride[1] * dst.elem_size;
+
+        // Just do array assignment if elements are small enough.
+        // It's slightly ridiculous that we have all these versions, but oh well.
```

Comment:
Could collapse into a local templated function if you want something a little terser.